dwave/preprocessing/composites/spin_reversal_transform.py

```diff
@@ -91,38 +133,52 @@ def sample(self, bqm, *, num_spin_reversal_transforms=1, **kwargs):
             >>> len(response)
             400
         """
-        if not bqm.num_variables:
-            return self.child.sample(bqm, **kwargs)
-
-        # make a main response
-        responses = []
-
-        flipped_bqm = bqm.copy()
-        transform = {v: False for v in bqm.variables}
-        
-        for ii in range(num_spin_reversal_transforms):
-            # flip each variable with a 50% chance
-            for v in bqm.variables:
-                if random() > .5:
-                    transform[v] = not transform[v]
-                    flipped_bqm.flip_variable(v)
-
-            flipped_response = self.child.sample(flipped_bqm, **kwargs)
-
-            tf_idxs = [flipped_response.variables.index(v)
-                       for v, flip in transform.items() if flip]
-
-            if bqm.vartype is Vartype.SPIN:
-                flipped_response.record.sample[:, tf_idxs] = -1 * flipped_response.record.sample[:, tf_idxs]
-            else:
-                flipped_response.record.sample[:, tf_idxs] = 1 - flipped_response.record.sample[:, tf_idxs]
-
-            if num_spin_reversal_transforms == 1:
-                #Case of a single sampleset call, avoid
-                #stripping of info field and other
-                #information. 
-                return(flipped_response)
-            else:
-                responses.append(flipped_response)
-
-        return concatenate(responses)
+        sampler = self._child
+
+        # No SRTs, so just pass the problem through
+        if not num_spin_reversal_transforms or not bqm.num_variables:
+            sampleset = sampler.sample(bqm, **kwargs)
+            # yield twice because we're using the @nonblocking_sample_method
+            yield sampleset  # this one signals done()-ness
+            yield sampleset  # this is the one actually used by the user
+            return
+
+        # we'll be modifying the BQM, so make a copy
+        bqm = bqm.copy()
+
+        # We maintain the Leap behavior that num_spin_reversal_transforms == 1
+        # corresponds to a single problem with randomly flipped variables.
+
+        # Get the SRT matrix
+        SRT = self.rng.random((num_spin_reversal_transforms, bqm.num_variables)) > .5
+
+        # Submit the problems
+        samplesets: typing.List[dimod.SampleSet] = []
+        flipped = np.zeros(bqm.num_variables, dtype=bool)  # what variables are currently flipped
+        for i in range(num_spin_reversal_transforms):
+            # determine what needs to be flipped
+            transform = flipped != SRT[i, :]
```

Comment:
Clever trick to minimize flips.

We could further optimize it by sorting the `SRT` by Hamming distance, or in "gray code order".